### PR TITLE
8255235: ZGC: Allocate and initialize forwarding data structures in parallel

### PR DESCRIPTION
**Pull request by pliden**
As a continuation of JDK-8255234, we can speed things up even more by allocating/initializing forwarding data structures in parallel. For example, testing using BigRamTester shows the time for "Concurrent Select Relocation Set" phase drop by ~50% when using 4 threads.

This PR depends on "8255234: ZGC: Bulk allocate forwarding data structures", which is a separate PR (https://github.com/openjdk/jdk/pull/804).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8255235](https://bugs.openjdk.java.net/browse/JDK-8255235): ZGC: Allocate and initialize forwarding data structures in parallel

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.java.net/census#ayang) (@albertnetymk - Author) ⚠️ Review applies to 6f340aa1965490c2fb13ff6f55272e5a0390420c
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/805/head:pull/805`
`$ git checkout pull/805`

### CODE REVIEW

File: src/hotspot/share/gc/z/zRelocationSet.cpp [hunk 1]
```diff
@@ -22,35 +22,102 @@
...
+  ZForwarding**                  _forwardings;
+  const size_t                   _nforwardings;
+  ZArrayParallelIterator<ZPage*> _small_iter;
+  ZArrayParallelIterator<ZPage*> _medium_iter;
+  volatile size_t                _small_next;
+  volatile size_t                _medium_next;
+
+  void install(ZForwarding* forwarding, volatile size_t* next) {
+    const size_t index = Atomic::fetch_and_add(next, 1u);
+    _forwardings[index] = forwarding;
```
Comment (albertnetymk) [inline src/hotspot/share/gc/z/zRelocationSet.cpp:47]: Maybe `assert(index < _nforwardings)` before indexing the array?

File: src/hotspot/share/gc/z/zRelocationSet.cpp [hunk 2]
```diff
@@ -22,35 +22,102 @@
...
+    _workers(workers),
+    _allocator(),
+    _forwardings(NULL),
+    _nforwardings(0) {}
+
+void ZRelocationSet::install(const ZRelocationSetSelector* selector) {
+  // Install relocation set
+  ZRelocationSetInstallTask task(&_allocator, selector);
+  _workers->run_concurrent(&task);
+
```
Comment (albertnetymk) [inline src/hotspot/share/gc/z/zRelocationSet.cpp:113]: Now all space in `_allocator` is used; I wonder if it's worth placing an assert here, sth like `assert(_allocator.used() == _allocator.size())`.

**Review by albertnetymk — APPROVED** [review]
Only minor comments on additional asserts.

File: src/hotspot/share/gc/z/zRelocationSet.cpp (hunk 1 above)
Comment (pliden) [reply src/hotspot/share/gc/z/zRelocationSet.cpp:47]: Will fix.

File: src/hotspot/share/gc/z/zRelocationSet.cpp (hunk 2 above)
Comment (pliden) [reply src/hotspot/share/gc/z/zRelocationSet.cpp:113]: Sounds like a good idea. I'll add that to #804.

**Comment by albertnetymk** [comment]
Thank you for the change.

**Review by fisk — APPROVED** [review]
Looks good.

**Pull request closed without merging**